presto-memory-context/src/main/java/com/facebook/presto/memory/SimpleLocalMemoryContext.java

```diff
@@ -0,0 +1,131 @@
+
+import static com.facebook.presto.memory.MemoryReservationHandler.NOT_BLOCKED;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class SimpleLocalMemoryContext
+        implements LocalMemoryContext
+{
+    @Nonnull
+    private final AggregatedMemoryContext parentMemoryContext;
+    @GuardedBy("this")
+    private long usedBytes;
+    @GuardedBy("this")
+    private boolean closed;
+
+    public SimpleLocalMemoryContext(AggregatedMemoryContext parentMemoryContext)
+    {
+        this.parentMemoryContext = requireNonNull(parentMemoryContext);
+    }
+
+    @Override
+    public synchronized long getBytes()
+    {
+        return usedBytes;
+    }
+
+    @Override
+    public synchronized ListenableFuture<?> setBytes(long bytes)
+    {
+        checkClosed();
+
+        if (bytes == usedBytes) {
+            return NOT_BLOCKED;
+        }
+
+        // update the parent first as it may throw a runtime exception (e.g., ExceededMemoryLimitException)
+        ListenableFuture<?> future = parentMemoryContext.updateBytes(bytes - usedBytes);
+        usedBytes = bytes;
+        return future;
+    }
+
+    @Override
+    public synchronized boolean tryUpdateBytes(long bytes)
+    {
+        checkClosed();
+
+        if (parentMemoryContext.tryUpdateBytes(bytes)) {
+            usedBytes += bytes;
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * This method transfers ownership of allocations from this memory context to the "to" memory context,
+     * where parent of this is a descendant of to.parent (there can be multiple AggregatedMemoryContexts between them).
+     *
+     * During the transfer we don't call the reservation handlers to reflect any state changes outside of the contexts.
+     */
+    @Override
+    public synchronized void transferOwnership(LocalMemoryContext to)
+    {
+        checkArgument(to instanceof SimpleLocalMemoryContext, "to has to be SimpleLocalMemoryContext");
+        checkClosed();
+
+        SimpleLocalMemoryContext target = (SimpleLocalMemoryContext) to;
+        AggregatedMemoryContext parent = parentMemoryContext;
+        while (parent != null && parent != target.parentMemoryContext) {
+            synchronized (parent) {
+                parent.usedBytes -= usedBytes;
+            }
+            parent = parent.parentMemoryContext;
+        }
+
+        synchronized (to) {
+            target.usedBytes += usedBytes;
+        }
+
+        usedBytes = 0;
+    }
+
+    private void checkClosed()
+    {
+        if (closed) {
+            throw new IllegalStateException("SimpleLocalMemoryContext is already closed");
+        }
+    }
+
+    @Override
+    public synchronized void close()
+    {
+        if (closed) {
+            return;
+        }
+        closed = true;
+        if (parentMemoryContext != null) {
```

Comment:
This `if` is redundant.

Question for you, @nezihyigitbasi: Do you see IntelliJ "warning"s for this? If not, you should turn those on.